if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager", lib = "/home/j/PRACTICA_RNASEQ/library/")
BiocManager::install("DESeq2", lib = "/home/j/PRACTICA_RNASEQ/library/")
BiocManager::install("readr", lib = "/home/j/PRACTICA_RNASEQ/library/")
BiocManager::install("AnnotationDbi", lib = "/home/j/PRACTICA_RNASEQ/library/")
BiocManager::install("org.Hs.eg.db", lib = "/home/j/PRACTICA_RNASEQ/library/")
BiocManager::install("ggbeeswarm", lib = "/home/j/PRACTICA_RNASEQ/library/")
BiocManager::install("clusterProfiler", lib = "/home/j/PRACTICA_RNASEQ/library/")
BiocManager::install("GenomicFeatures", lib = "/home/j/PRACTICA_RNASEQ/library/")

BiocManager::install("tximeta", lib = "/home/j/PRACTICA_RNASEQ/library/")
BiocManager::install("ReportingTools", lib = "/home/j/PRACTICA_RNASEQ/library/")
install.packages(c("pheatmap","RColorBrewer","ggplot2"), lib = "/home/j/PRACTICA_RNASEQ/library/")

library(DESeq2)
library(readr)
library(AnnotationDbi)
library(org.Hs.eg.db, lib.loc = "/home/j/PRACTICA_RNASEQ/library/")
#Set your working directoty to the folder RNASeqPractical_2
library(ggbeeswarm, lib.loc = "/home/j/PRACTICA_RNASEQ/library/")
#library(clusterProfiler, lib.loc = "/home/j/PRACTICA_RNASEQ/library/")
library(tximeta, lib.loc = "/home/j/PRACTICA_RNASEQ/library/")
#library(ReportingTools, lib.loc = "/home/j/PRACTICA_RNASEQ/library/")
library(pheatmap, lib.loc = "/home/j/PRACTICA_RNASEQ/library/")
library(RColorBrewer, lib.loc = "/home/j/PRACTICA_RNASEQ/library/")
library(ggplot2 , lib.loc =  "/home/j/PRACTICA_RNASEQ/library/")
library(GenomicFeatures , lib.loc =  "/home/j/PRACTICA_RNASEQ/library/")

setwd("/home/j/PRACTICA_RNASEQ/")
#Load the information about the experiment, sample names, and condition
targets<-read.table("metadata.csv",
                    header=T,
                    sep = ",",
                    stringsAsFactors=FALSE,
                    col.names = c("names", "condition", "cell"))
targets$name<-as.factor(targets$name)
targets$cell<-as.factor(targets$cell)
targets$group <- as.factor(paste0(targets$condition, "_", targets$cell))

#targets$condition <- relevel(targets$condition, "Untreated")

#make a list of the quantification files created by salmon. 
#Please note these are only strings
files<-file.path(paste("Quantification/",targets$names,"_salmon/quant.sf",sep=''))

#check that the string (file names) created in the previous steps are
#actual files on disk
file.exists(files)

#Add the file names to the description of  experiment.
targets$files<-files
row.names(targets)<-targets$names
#check the targets object
targets

#We will import the transcript quantification generated by salmon,
#and the experiment description using tximeta
#Check the manual of tximeta for further details:
#https://bioconductor.org/packages/3.14/bioc/vignettes/tximeta/inst/doc/tximeta.html

#First load the tximeta package
library("tximeta")
install.packages("tximeta")

#Now, before loading our data, create tthe transcriptome object,
#this will help to carry out several operations,
#like summarizing the transcript expression levels at the gene level,
#get functional information about the genes, and so on.

makeLinkedTxome(indexDir="/home/j/PRACTICA_RNASEQ/salmon/salmon_index/",
                source="LocalEnsembl",
                organism="Homo sapiens",
                release="38",
                genome="GRCh38.p13",
                fasta="/home/j/PRACTICA_RNASEQ/salmon/gencode.v38.transcripts.fa.gz",
                gtf="/home/j/PRACTICA_RNASEQ/salmon/gencode.v38.annotation.gtf.gz",
                write=FALSE)

#Now create the tximeta object that has the salmon data, the experiment
#description, and access to the transcriptome information. This object is called
#SummarizedExperiment 
se <- tximeta(targets)
#Check the size of the object
dim(se)
#check the names of the transcripts, confirm that you are seeing the transcript names
#and not gene names. How woudl you now.
head(rownames(se))

#Nos exploit the transcriptome information to summarize the expression data
#at the gene level. Do you understand what this means?
gse <- summarizeToGene(se)

#Check the size of the object and check the names of the genes. Should be now 
#gene names and not transcripts names
dim(gse)
head(rownames(gse))

#these two object are compossed of colData, ranges and assay data, see figure
#let's see each of these elements in our objects se and gse
#See the figure in section 2.5 of https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html

library(SummarizedExperiment)
colData(gse)
head(assay(gse))
rowRanges(gse)

#With these object (gse) we can now start the differential gene expression analysis
#For that there are many different R packages that could be used
#today we are going to use the DESeq2 package.
#So, first we need to transform our SummarizedExperiment (gse) object to an object
#that is native to DESeq2, which is DESeqDataSet, in addition to the data, 
#the DESeqDataSet requires a description of the experimental design,
#in which we tell what are the factor of interest in the study, and/or the
#factors that should be sources of variation and how to deal with them.
#This is done using a formula, with the same sintax as in simple linear models (lm) in R.

#In this particular example, we have at least (that we know of) sources
#of variation. First, the condition, samples treated or not with Dexamethasone.
#Second, the cell line. The primary ASM cell were obtained from four donors,
# and then divided in the treated and untreated groups. So we can have an effect of the donor.
#Check you targets object to see this.
#We will use these two factors in our analyses. We want to test the effect
#of Dexamethasone, while controlling for the effect of different cellLines.
#Note that this is a paired experimental design. As each sample (cellLine)
#was treated and not treated
#in that case we will specify the formula as ~ cellLine + condition

gse$cell
gse$condition

targets

library(DESeq2)
dds <- DESeqDataSet(gse, design = ~condition+cell)

#It is common that genes in the dataset at not expressed at all, and others
#to be very lowly expressed, it is a good practice to remove these from further analyses

nrow(dds) #Number of genes pre-filtering

keep <- rowSums(counts(dds)) > 1
dds <- dds[keep,]
nrow(dds) #Number of genes after filtering out non-expressed genes.

#Check the effect of normalization
head(counts(dds),2)
dds <- estimateSizeFactors(dds)
head(counts(dds,normalized=TRUE),2)
colSums(assay(gse))
#It is also a good idea to remove genes in which a given number of samples do not have an count

keep <- rowSums(counts(dds) >= 5) >= 4
dds <- dds[keep,]
nrow(dds) # Number of genes, that have counts in at least 4 samples, and the sum of counts is greated than 5


#Now that the data is in the proper format and filtered 
#we can carry out some exploratory analyses to check 
#whether the data looks good

#Many os the exploratory analysis require homocedastic data, which our counts are not.
#In DESeq2 the function vst and rlog, can normalize the data to make it more
#homocedastic. As a rule of thum vst is better when you have more than 30 samples

rld <- rlog(dds, blind = TRUE)
head(assay(rld))

#Now let's look at the similarity among samples
#we will compute the euclidean distance between the samples with the function dist
#This function assumes that the samples are the rows, so we must first transpose the matrix
#We will visualize the distances between the sampleas as a heatmap

library("pheatmap")
library("RColorBrewer")

sampleDists <- dist(t(assay(rld)))
sampleDists

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( rld$condition, rld$cell, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- inferno(50)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)

sampleDists <- dist(t(assay(rld)))
sampleDists

samplecor <- as.dist((1-(cor(assay(rld), method='pearson')))*100)
sampleCorMatrix <- as.matrix( samplecor )
rownames(sampleCorMatrix) <- paste( rld$condition, rld$cell, sep = " - " )
colnames(sampleCorMatrix) <- NULL
#colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleCorMatrix,
         clustering_distance_rows = samplecor,
         clustering_distance_cols = samplecor,
         col=colors)

#It is also common to show a PCA plot of the samples, colored with the factors of interest

pcaData <- plotPCA(rld, intgroup = c( "condition",
                                      "cell"), 
                   returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
library(ggplot2)
ggplot(pcaData, aes(x = PC1, y = PC2, color = condition, shape = cell)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA with RLOG data")

#These exploratory analyses can help us detect outliers, and rogue factors (other batch effects)
#Like technician, time of sample collection/preparation and so on.

dds <- DESeq(dds)
res <- results(dds)
par(las = 2)
plotCounts(dds, gene = "ENSG00000270872.2", intgroup = c("cell", "condition"))
